kimchi/src/circuits/cairo/runner/memory.rs

```diff
@@ -0,0 +1,137 @@
+//! This module represents the Cairo memory, containing the
+//! compiled Cairo program that occupies the first few entries
+
+use std::ops::{Index, IndexMut};
+
+use crate::circuits::cairo::runner::helper::*;
+use crate::circuits::cairo::runner::word::CairoWord;
+use ark_ff::PrimeField;
+use core::iter::repeat;
+use o1_utils::FieldHelpers;
+
+/// This data structure stores the memory of the program
+pub struct CairoMemory<F: PrimeField> {
+    /// length of the public memory
+    codelen: usize,
+    /// full memory vector, None if non initialized
+    pub data: Vec<Option<CairoWord<F>>>,
+}
+
+impl<F: PrimeField> Index<F> for CairoMemory<F> {
+    type Output = Option<CairoWord<F>>;
+    fn index(&self, idx: F) -> &Self::Output {
+        // Safely convert idx from F to usize (since this is a memory address
+        // idx should not be too big, this should be safe)
+        let addr: u64 = idx.to_u64();
+        &self.data[addr as usize]
```

Comment:
So it seems like I can call `resize` inside `index_mut()` to avoid having to call `resize()` inside `write()`, but I cannot avoid calling `resize()` in `read()`. Since `write()` is assigning a value to an address of memory, it looks like it is using the `IndexMut` trait beneath. Unlike `read()`, which is only "looking into" an address of memory, and thus seems to be calling `Index` instead. And since in `Index` the `self` is immutable, I cannot call `resize()` there.